pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -0,0 +1,791 @@
+	if !pl.enabled {
+		return nil
+	}
+	state, err := getStateData(cs)
+	if err != nil {
+		return statusError(klog.FromContext(ctx), err)
+	}
+	if len(state.claims) == 0 {
+		return nil
+	}
+
+	logger := klog.FromContext(ctx)
+	podScheduling, err := state.initializePodScheduling(ctx, pod, pl.podSchedulingLister)
+	if err != nil {
+		return statusError(logger, err)
+	}
+	pending := false
+	for _, claim := range state.claims {
+		if claim.Status.Allocation == nil {
+			pending = true
+		}
+	}
+	if pending && !haveAllNodes(podScheduling.Spec.PotentialNodes, nodes) {
+		// Remember the potential nodes. The object will get created or
+		// updated in Reserve. This is both an optimization and
+		// covers the case that PreScore doesn't get called when there
+		// is only a single node.
+		logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+		podScheduling = podScheduling.DeepCopy()
+		numNodes := len(nodes)
+		if numNodes > resourcev1alpha1.PodSchedulingNodeListMaxSize {
+			numNodes = resourcev1alpha1.PodSchedulingNodeListMaxSize
+		}
+		podScheduling.Spec.PotentialNodes = make([]string, 0, numNodes)
+		if numNodes == len(nodes) {
+			// Copy all node names.
+			for _, node := range nodes {
+				podScheduling.Spec.PotentialNodes = append(podScheduling.Spec.PotentialNodes, node.Name)
+			}
+		} else {
+			// Select a random subset of the nodes to comply with
+			// the PotentialNodes length limit. Randomization is
+			// done for us by Go which iterates over map entries
+			// randomly.
+			nodeNames := map[string]struct{}{}
+			for _, node := range nodes {
+				nodeNames[node.Name] = struct{}{}
+			}
+			for nodeName := range nodeNames {
+				if len(podScheduling.Spec.PotentialNodes) >= resourcev1alpha1.PodSchedulingNodeListMaxSize {
+					break
+				}
+				podScheduling.Spec.PotentialNodes = append(podScheduling.Spec.PotentialNodes, nodeName)
+			}
+		}
+		sort.Strings(podScheduling.Spec.PotentialNodes)
+		state.storePodScheduling(podScheduling)
+	}
+	logger.V(5).Info("all potential nodes already set", "pod", klog.KObj(pod), "potentialnodes", nodes)
+	return nil
+}
+
+func haveAllNodes(nodeNames []string, nodes []*v1.Node) bool {
+	for _, node := range nodes {
+		if !haveNode(nodeNames, node.Name) {
+			return false
+		}
+	}
+	return true
+}
+
+func haveNode(nodeNames []string, nodeName string) bool {
+	for _, n := range nodeNames {
+		if n == nodeName {
+			return true
+		}
+	}
+	return false
+}
+
+// Reserve reserves claims for the pod.
+func (pl *dynamicResources) Reserve(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
+	if !pl.enabled {
+		return nil
+	}
+	state, err := getStateData(cs)
+	if err != nil {
+		return statusError(klog.FromContext(ctx), err)
+	}
+	if len(state.claims) == 0 {
+		return nil
+	}
+
+	pending := 0
+	infos := 0
```

Comment:
what is this? Probably needs a better name.

Reply:
```
	numDelayedAllocationPending := 0
	numClaimsWithStatusInfo := 0
```